### PR TITLE
ARROW-3331: [Gandiva][C++] Add re2 to toolchain

### PR DESCRIPTION
**Pull request by pravindra**

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1057,11 +1063,47 @@ if (ARROW_WITH_ZSTD)
   endif()
 endif()
 
+# ----------------------------------------------------------------------
+# RE2 (required for Gandiva)
+if (ARROW_GANDIVA)
+  # re2
+  if ("${RE2_HOME}" STREQUAL "")
```
Comment (wesm) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: This also needs to be added to the `$ARROW_BUILD_TOOLCHAIN` part at the top of the file. I will add and push here

**Review by wesm — APPROVED** [review]
Minor change re the `$ARROW_BUILD_TOOLCHAIN` which is used in Travis CI and Python builds. I will test this out locally, add that, and then merge

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (wesm) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: This was already there. This is failing for me for another reason so I'll push a fix

**Comment by wesm** [comment]
This appears to have a race condition, encountered when running `make -j8`. Looking into it

```
[ 21%] Building CXX object src/gandiva/CMakeFiles/gandiva_obj_lib.dir/bitmap_accumulator.cc.o
[ 22%] Building CXX object src/gandiva/CMakeFiles/gandiva_obj_lib.dir/configuration.cc.o
[ 22%] Building CXX object src/gandiva/CMakeFiles/gandiva_obj_lib.dir/engine.cc.o
[ 22%] Building CXX object src/gandiva/CMakeFiles/gandiva_obj_lib.dir/expr_decomposer.cc.o
In file included from /home/wesm/code/arrow/cpp/src/gandiva/function_holder_registry.h:26:0,
                 from /home/wesm/code/arrow/cpp/src/gandiva/expr_decomposer.cc:27:
/home/wesm/code/arrow/cpp/src/gandiva/like_holder.h:24:21: fatal error: re2/re2.h: No such file or directory
 #include <re2/re2.h>
                     ^
compilation terminated.
-- glog_ep configure command succeeded.  See also /home/wesm/code/arrow/cpp/fresh-build/glog_ep-prefix/src/glog_ep-stamp/glog_ep-configure-*.log
```

**Comment by wesm** [comment]
Still some more dependency issues:

```
$ make -j8 epoch_time_point_test
Scanning dependencies of target epoch_time_point_test
make[3]: *** No rule to make target `googletest_ep-prefix/src/googletest_ep/lib/libgtest.a', needed by `debug/epoch_time_point_test'.  Stop.
make[3]: *** Waiting for unfinished jobs....
Building CXX object src/gandiva/precompiled/CMakeFiles/epoch_time_point_test.dir/epoch_time_point_test.cc.o
/home/wesm/code/arrow/cpp/src/gandiva/precompiled/epoch_time_point_test.cc:20:25: fatal error: gtest/gtest.h: No such file or directory
 #include <gtest/gtest.h>
                         ^
compilation terminated.
make[3]: *** [src/gandiva/precompiled/CMakeFiles/epoch_time_point_test.dir/epoch_time_point_test.cc.o] Error 1
make[2]: *** [src/gandiva/precompiled/CMakeFiles/epoch_time_point_test.dir/all] Error 2
make[1]: *** [src/gandiva/precompiled/CMakeFiles/epoch_time_point_test.dir/rule] Error 2
make: *** [epoch_time_point_test] Error 2
```

```
$ make -j8 expression_registry_test
<SNIP>
Scanning dependencies of target arrow_shared
[ 94%] Linking CXX shared library ../../debug/libarrow.so
[ 94%] Built target arrow_shared
Scanning dependencies of target expression_registry_test
[ 97%] Building CXX object src/gandiva/CMakeFiles/expression_registry_test.dir/llvm_types.cc.o
[ 97%] Building CXX object src/gandiva/CMakeFiles/expression_registry_test.dir/expression_registry_test.cc.o
make[3]: *** No rule to make target `googletest_ep-prefix/src/googletest_ep/lib/libgtest.a', needed by `debug/expression_registry_test'.  Stop.
make[3]: *** Waiting for unfinished jobs....
[ 97%] Building CXX object src/gandiva/CMakeFiles/expression_registry_test.dir/expression_registry.cc.o
[ 97%] Building CXX object src/gandiva/CMakeFiles/expression_registry_test.dir/function_signature.cc.o
[100%] Building CXX object src/gandiva/CMakeFiles/expression_registry_test.dir/function_registry.cc.o
/home/wesm/code/arrow/cpp/src/gandiva/expression_registry_test.cc:23:25: fatal error: gtest/gtest.h: No such file or directory
 #include <gtest/gtest.h>
                         ^
compilation terminated.
make[3]: *** [src/gandiva/CMakeFiles/expression_registry_test.dir/expression_registry_test.cc.o] Error 1
make[2]: *** [src/gandiva/CMakeFiles/expression_registry_test.dir/all] Error 2
make[1]: *** [src/gandiva/CMakeFiles/expression_registry_test.dir/rule] Error 2
make: *** [expression_registry_test] Error 2
```

Ideally we should replace all usages of `add_gandiva_unit_test` with `ADD_ARROW_TEST` as it handles the toolchain dependencies etc.

**Comment by wesm** [comment]
This looks good new. Let me know if my changes are OK and then I'll merge. I'm getting a weird error building on Ubuntu 14.04 that's unrelated to this:

```
[ 90%] Linking CXX executable ../../../debug/hash_test_gandiva_static
/usr/lib/llvm-6.0/lib/libLLVMX86CodeGen.a(X86FloatingPoint.cpp.o): In function `(anonymous namespace)::FPS::runOnMachineFunction(llvm::MachineFunction&) [clone .part.165]':
(.text._ZN12_GLOBAL__N_13FPS20runOnMachineFunctionERN4llvm15MachineFunctionE.part.165+0x481): undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)'
/usr/lib/llvm-6.0/lib/libLLVMX86CodeGen.a(X86InstructionSelector.cpp.o): In function `llvm::PredicateBitsetImpl<107ul>::PredicateBitsetImpl(std::initializer_list<unsigned int>)':
(.text._ZN4llvm19PredicateBitsetImplILm107EEC2ESt16initializer_listIjE[_ZN4llvm19PredicateBitsetImplILm107EEC5ESt16initializer_listIjE]+0x75): undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)'
/usr/lib/llvm-6.0/lib/libLLVMAsmPrinter.a(CodeViewDebug.cpp.o): In function `llvm::CodeViewDebug::getFullFilepath(llvm::DIFile const*)':
(.text._ZN4llvm13CodeViewDebug15getFullFilepathEPKNS_6DIFileE+0x452): undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)'
/usr/lib/llvm-6.0/lib/libLLVMAsmPrinter.a(CodeViewDebug.cpp.o): In function `llvm::CodeViewDebug::getFullFilepath(llvm::DIFile const*)':
(.text._ZN4llvm13CodeViewDebug15getFullFilepathEPKNS_6DIFileE+0x46a): undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)'
/usr/lib/llvm-6.0/lib/libLLVMCodeGen.a(TargetLoweringBase.cpp.o): In function `getOpEnabled(bool, llvm::EVT, llvm::StringRef)':
(.text._ZL12getOpEnabledbN4llvm3EVTENS_9StringRefE+0x3f3): undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)'
/usr/lib/llvm-6.0/lib/libLLVMCodeGen.a(TargetLoweringBase.cpp.o):(.text._ZL20getOpRefinementStepsbN4llvm3EVTENS_9StringRefE+0x367): more undefined references to `std::__throw_out_of_range_fmt(char const*, ...)' follow
collect2: error: ld returned 1 exit status
make[2]: *** [debug/hash_test_gandiva_static] Error 1
make[1]: *** [src/gandiva/tests/CMakeFiles/hash_test_gandiva_static.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by wesm** [comment]
I think the reason I'm getting that error is because LLVM is compiled with a newer libstdc++ (for LLVM 6 this is required on Ubuntu 14.04), but since I'm compiling with gcc 4.8.x the flags `-static-libstdc++ -static-libgcc` are causing a conflict

**Comment by pravindra** [comment]
@wesm Your changes look good. I ran on ubuntu 14.04 with gcc-4.9 and so, didn't see the LLVM linking error.

**Comment by wesm** [comment]
OK. I think the `static-libgcc static-libstdc++` stuff should be handled by LDFLAGS rather than hardcoded. I can open a JIRA about it?

**Comment by pravindra** [comment]
> OK. I think the `static-libgcc static-libstdc++` stuff should be handled by LDFLAGS rather than hardcoded. I can open a JIRA about it?

sure, pls do.

**Comment by wesm** [comment]
Cool. https://issues.apache.org/jira/browse/ARROW-3437. I'll rebase this patch and then merge it

**Pull request closed without merging**